python/ray/serve/api.py

```diff
@@ -1451,3 +1517,381 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
+    def _add_deployment(self, deployment: Deployment):
+        """Adds the deployment to this Serve application.
+
+        Validates that a deployment with the same name doesn't already exist.
+        To overwrite an existing deployment, use index notation. For example:
+
+        app[deployment_name] = deployment
+
+        Args:
+            deployment (Deployment): deployment to add to this Application.
+
+        Raises:
+            ValueError: If a deployment with deployment.name already exists in
+                this Application.
+        """
+
+        if not isinstance(deployment, Deployment):
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+        elif deployment.name in self._deployments:
+            raise ValueError(
+                f'Deployment with name "{deployment.name}" already '
+                "exists in this application. To overwrite this "
+                "deployment, use attribute or index notation. "
+                "For example:\n\napp.deployment_name = "
+                "new_deployment"
+            )
+
+        self._deployments[deployment.name] = deployment
+
+    def _get_deployments(self) -> List[Deployment]:
+        """Gets a list of this Application's deployments."""
+
+        return self._deployments.values()
+
+    def __getitem__(self, key: str):
+        """Fetch a deployment by name using dict syntax: app["name"]
+
+        Raises:
+            KeyError: if the name is not in this Application.
+        """
+
+        if key in self._deployments:
+            return self._deployments[key]
+        else:
+            raise KeyError(f'Serve application does not contain a "{key}" deployment.')
+
+    def __iter__(self):
+        """Iterator over Application's deployments.
+
+        Enables "for deployment in Application" pattern.
+        """
+
+        return iter(self._deployments.values())
+
+    def __len__(self):
+        """Number of deployments in this Application."""
+
+        return len(self._deployments)
+
+    def __contains__(self, key: str):
+        """Checks if the key exists in self._deployments."""
+
+        return key in self._deployments
+
+
+@PublicAPI(stability="alpha")
+def run(
+    target: Union[DeploymentNode, Application],
+    *,
+    http_options: Union[Dict, HTTPOptions] = None,
+    logger: Optional[logging.Logger] = None,
+):
+    """Run a Serve application (blocking).
+
+    Deploys all of the deployments in the application then blocks,
+    periodically logging the status. Upon a KeyboardInterrupt, the application
+    will be torn down and all of its deployments deleted.
+
+    Either a DeploymentNode or a pre-built application can be passed in.
+    If a DeploymentNode is passed in, all of the deployments it depends on
+    will be deployed.
+    """
+
+    try:
+        deployments = _get_deployments_from_target(target)
+
+        # TODO (shrekris-anyscale): validate ingress
+
+        start(detached=True, http_options=http_options)
```

Comment:
can we log a good error message if serve is already started & the http_options don't match? cc @simon-mo do we do this already?

Reply:
It looks like this already happens [here](https://github.com/ray-project/ray/blob/022068b1ad3db6d93704d514885532112b5869a2/python/ray/serve/api.py#L688).

Reply:
excelltn